src/test/java/software/amazon/awssdk/crt/test/DirectoryTraversalTest.java

```diff
@@ -0,0 +1,250 @@
+
+        try (final DirectoryStructureHelper directoryStructure = new DirectoryStructureHelper()) {
+            // traverse dir using CRT
+            Set<String> entries = new HashSet<>();
+
+            try {
+                DirectoryTraversal.traverse(directoryStructure.getRootDirectory(), false, new DirectoryTraversalHandler() {
+                    @Override
+                    public boolean onDirectoryEntry(final DirectoryEntry directoryEntry) {
+                        entries.add(directoryEntry.getPath());
+                        assertTrue(directoryEntry.isDirectory());
+                        assertTrue(!directoryEntry.isSymLink());
+                        assertTrue(!directoryEntry.isFile());
+
+                        return false;
+                    }
+                });
+
+                assertTrue("Cancellation should have caused an exception", false);
+            } catch (final RuntimeException ex) {
+                // should return only one entry, because the callback returned false asking the traversal to be cancelled.
+                assertEquals(1, entries.size());
+            }
+        }
+    }
+
+    @Test
+    public void testTraverseDirectoryError() throws Exception {
+
+        Set<String> entries = new HashSet<>();
+
+        try {
+
+            DirectoryTraversal.traverse("-", false, new DirectoryTraversalHandler() {
+                @Override
+                public boolean onDirectoryEntry(final DirectoryEntry directoryEntry) {
+                    entries.add(directoryEntry.getPath());
+                    return true;
+                }
+            });
+            assertTrue("Cancellation should have caused an exception", false);
+        } catch (final RuntimeException ex) {
+            assertEquals(0, entries.size());
+        }
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testTraverseDirectoryNullPath() {
+
+        DirectoryTraversal.traverse(null, false, new DirectoryTraversalHandler() {
+            @Override
+            public boolean onDirectoryEntry(final DirectoryEntry directoryEntry) {
+                return true;
+            }
+        });
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void testTraverseDirectoryNullHandler() {
+
+        DirectoryTraversal.traverse("-", false, null);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void testTraverseDirectoryCallbackThrowingException() throws Exception {
+
+        try (final DirectoryStructureHelper directoryStructure = new DirectoryStructureHelper()) {
+            // traverse dir using CRT
+            Set<String> entries = new HashSet<>();
+
+            DirectoryTraversal.traverse(directoryStructure.getRootDirectory(), false, new DirectoryTraversalHandler() {
+                @Override
+                public boolean onDirectoryEntry(final DirectoryEntry directoryEntry) {
+                    throw new RuntimeException("THROWN BY TEST CALLBACK");
```

Comment:
with the existing code, what happens to `RuntimeException("THROWN BY TEST CALLBACK")`?
is it:
1) `RuntimeException("THROWN BY TEST CALLBACK")` is raised?
2) `RuntimeException("AWS_ERROR_OPERATION_INTERUPTED")` is raised?
3) `RuntimeException("AWS_ERROR_OPERATION_INTERUPTED")` which wraps `RuntimeException("THROWN BY TEST CALLBACK")` is raised?

I know I'd personally prefer 1 or 3, so I can see the actual reason for the failure

Reply:
The current behavior is 1. I just added an assertion in the unit test for this particular check. Thanks!